### PR TITLE
Add Perceptron binary classifier

### LINKED ISSUE #7599: [FEATURE REQUEST] Add Perceptron binary classifier
## What would you like to Propose?

Add a Perceptron classifier to `src/main/java/com/thealgorithms/machinelearning`. The Perceptron is the simplest neural network and would provide an educational binary linear-classification algorithm alongside the existing `LinearRegression`, `KNearestNeighbors`, and `MultinomialNaiveBayesClassifier` implementations.

## Issue details

**Algorithm name:** Perceptron

**Problem statement:** Given a set of feature vectors and binary class labels, learn a linear decision boundary using the Perceptron learning rule, then classify previously unseen samples. The implementation should make the bias term explicit and document that convergence is guaranteed only for linearly separable data.

**Suggested scope:**
- Add `Perceptron.java` in the machinelearning package.
- Use only the Java standard library; no external machine-learning dependency is needed.
- Provide a small, clear API for fitting, predicting one sample, and predicting a batch.
- Support configurable learning rate and maximum epochs, with deterministic zero-weight and zero-bias initialization.
- Validate null or empty data, inconsistent feature dimensions, invalid labels, and invalid hyperparameters with clear exceptions.
- Document the update rule, label convention, and limitations in Javadoc.

**Acceptance tests:**
- Train on a linearly separable toy dataset such as AND or OR and classify all training samples correctly.
- Verify predictions for unseen samples and batch prediction.
- Verify that prediction before fitting fails clearly.
- Verify invalid labels, mismatched dimensions, null or empty input, and invalid learning-rate or epoch values.
- Include a non-separable-data test that checks documented behavior, such as stopping after the epoch limit without claiming convergence.

A historical pull request, [#187](https://github.com/TheAlgorithms/Java/pull/187), attempted a Perceptron implementation in 2018, but there is no current implementation in the package. This request is for a current Java 21 implementation with tests that follow the repository conventions.

### PR DESCRIPTION
**Pull request by Th-Shivam**
## Summary

Adds a binary Perceptron classifier to `com.thealgorithms.machinelearning`. The implementation uses deterministic zero initialization, an explicit bias term, configurable learning rate and epoch limit, input validation, single and batch prediction, defensive weight access, and convergence metadata.

## Tests

- AND toy dataset converges and predicts training samples correctly.
- Unseen and batch predictions are covered.
- XOR non-separable data stops at the configured epoch limit without reporting convergence.
- Re-fitting, defensive copies, pre-fit access, invalid hyperparameters, invalid training data, and invalid prediction data are covered.

## Verification

- Java 21-compatible compilation passed.
- Direct JUnit Platform execution: 11 tests passed.
- `git diff --check` passed.
- Full Maven/CI checks could not be run in this environment because Maven and clang-format are not installed; upstream CI should run them.

## Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Java/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized it.
- [x] All filenames are in PascalCase.
- [x] All functions and variable names follow Java naming conventions.
- [x] The new algorithm has a Wikipedia URL in its comments.
- [x] The new algorithm includes a corresponding test class.
- [x] New code follows the repository formatting style.

Closes #7599

### CODE REVIEW

File: src/test/java/com/thealgorithms/machinelearning/PerceptronTest.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
+package com.thealgorithms.machinelearning;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
```
Comment (alxkm) [inline src/test/java/com/thealgorithms/machinelearning/PerceptronTest.java]: This is what's failing the build. PMD's `TooManyStaticImports` allows at most 4 static imports per file; there are 5 here.

Import `Assertions` once and qualify the calls, as the rest of the test suite does:

```suggestion
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
```

Then update the call sites:

```diff
-        assertArrayEquals(labels, perceptron.predict(features));
-        assertTrue(perceptron.hasConverged());
+        Assertions.assertArrayEquals(labels, perceptron.predict(features));
+        Assertions.assertTrue(perceptron.hasConverged());
```

A wildcard static import would silence this rule but star imports are rejected too. Verify

**Review by alxkm — CHANGES_REQUESTED** [review]
Implementation looks good overall - validation runs before any mutation in `fit()`, `getWeights()` returns a defensive copy, and there's no `Random`, so the tests are deterministic. I traced AND and XOR by hand and the convergence behaviour matches what the tests assert.

Two things:

**1. PMD is failing the build.** `TooManyStaticImports` allows at most 4 static imports per file. Import `Assertions` once and qualify the calls, as the rest of the test suite does:

```java
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
```

```diff
-        assertArrayEquals(labels, perceptron.predict(features));
+        Assertions.assertArrayEquals(labels, perceptron.predict(features));
```

A wildcard static import would silence the rule but star imports are rejected too. `mvn -B clean verify` reproduces it locally.

**2. `getBias()` has no positive-path coverage.** It's only reached through the not-fitted path, so `return bias;` never executes. `getWeights()` is similar - the copy semantics are tested, the learned values aren't. One test covers both:

```java
@Test
void learnsExpectedWeightsAndBias() {
    Perceptron perceptron = new Perceptron(1.0, 10);
    perceptron.fit(new double[][] {{0}, {1}}, new int[] {0, 1});

    Assertions.assertArrayEquals(new double[] {1.0}, perceptron.getWeights());
    Assertions.assertEquals(-1.0, perceptron.getBias());
}
```

The model converges on epoch 3 with exact values, so no delta is needed.

**Comment by alxkm** [comment]
Test compilation is failing with `cannot find symbol` on every assertion. The `Assertions.` prefix needs to go on all the call sites, not just the import block - `assertArrayEquals`, `assertTrue`, `assertEquals`, `assertFalse`, `assertThrows` throughout the file. There's also a duplicated `import org.junit.jupiter.api.Test;` to drop.

Static imports can't come back: five distinct assertions, and PMD's `TooManyStaticImports` allows four. `mvn -B clean verify` to confirm before push.
